### PR TITLE
St/mempool nonce gap

### PR DESCRIPTION
**Pull request by saratnt**

### CODE REVIEW

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 1]
```diff
@@ -228,53 +238,79 @@ class MempoolMap(
...
+            destMap.put(tx.getNonce, tx.id)
+          } else {
+            if (!haveBecomeNonExecutable) {
+              destMap = newNonExecTxs
+              haveBecomeNonExecutable = true
+            }
           }
         }
+        else if (!maxNonceGapExceeded)
+          maxNonceGapExceeded = true
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:285]:
```suggestion
        else
          maxNonceGapExceeded = true
```
we may rely on `.withFilter(_ => !maxNonceGapExceeded)` so, as soon we reached this case we will never get inside this part.

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 2]
```diff
@@ -309,58 +360,46 @@ class MempoolMap(
...
+      //The new expected nonce is at least the same as the old one, so I don't need to check the remaining txs
+      //for the nonce gap. No need to check for tx size too, because this check was already performed when they were added
+      // the first time in the mempool
+      val fromAddress = account.asInstanceOf[AddressProposition].address()
+      val balance = accountStateReaderProvider.getAccountStateReader().getBalance(fromAddress)
+
+      val newExecTxs: mutable.TreeMap[BigInteger, ModifierId] = new mutable.TreeMap[BigInteger, ModifierId]()
+      val newNonExecTxs: mutable.TreeMap[BigInteger, ModifierId] = new mutable.TreeMap[BigInteger, ModifierId]()
+
+      // Recreates from scratch the account's nonExecTxs and execTxs maps, starting from txs from rejected blocks.
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:382]: this comment seems outdated.
Here we don't have any rejected txs, but only re-validate the existing mempool txs

File: sdk/src/main/scala/com/horizen/account/mempool/exception/MempoolException.scala [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.horizen.account.mempool.exception
+import sparkz.util.ModifierId
+
+import java.math.BigInteger
+
+class MempoolException(message: String) extends Exception(message)
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/account/mempool/exception/MempoolException.scala:6]: Do we want to move `TxOversizedException` here as well?

File: sdk/src/main/scala/com/horizen/certificatesubmitter/AbstractCertificateSubmitter.scala [hunk 4]
```diff
@@ -82,6 +82,7 @@ abstract class AbstractCertificateSubmitter[
   protected val certificateFee: Option[String] = if (settings.withdrawalEpochCertificateSettings.certificateAutomaticFeeComputation) None else Some(settings.withdrawalEpochCertificateSettings.certificateFee)
 
   override def preStart(): Unit = {
+    log.error("******************** AbstractCertificateSubmitter preStart")
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/certificatesubmitter/AbstractCertificateSubmitter.scala:85]: this line can be removed

File: sdk/src/test/scala/com/horizen/account/performance/AccountSidechainNodeViewHolderPerfTest.scala [hunk 5]
```diff
@@ -113,6 +113,163 @@ class AccountSidechainNodeViewHolderPerfTest
...
+    }
+  }
+
+  /*
+    Same as txModifyTest but it takes into account the nonce gap. So the maximum number of txs for each account can be
+    only 16. The other test is temporarily kept just for comparing the performances with the previous implementations
+    of the mempool.
+   */
+  @Test
+  //@Ignore
```
Comment (i-Alex) [inline sdk/src/test/scala/com/horizen/account/performance/AccountSidechainNodeViewHolderPerfTest.scala:241]: don't forget to move all the performance tests to `Ignore` again.

**Review by i-Alex — CHANGES_REQUESTED** [review]
Good!
I appreciate a lot the updated performance test and the comparison with/without nonce limits.

I share your doubt about the fork (especially MC fork) impact on txs returned back to mempool. With the default nonce gap limit of `16` we may loose a lot of transactions.
It will be considered in mc forks related topic.

File: sdk/src/main/scala/com/horizen/account/AccountSidechainApp.scala [hunk 6]
```diff
@@ -140,6 +140,7 @@ class AccountSidechainApp @Inject()
   val sidechainBlockActorRef: ActorRef = SidechainBlockActorRef[PMOD, SidechainSyncInfo, AccountHistory]("AccountBlock", sidechainSettings, sidechainBlockForgerActorRef)
 
   // Init Certificate Submitter
+  log.error("******************** before AccountCertificateSubmitterRef")
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/AccountSidechainApp.scala:143]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 7]
```diff
@@ -43,10 +47,16 @@ class MempoolMap(
         throw TxOversizedException(account.asInstanceOf[AddressProposition].address(), ethTransaction.size())
       }
 
+      val stateNonce = accountStateReaderProvider.getAccountStateReader().getNonce(account.asInstanceOf[AddressProposition].address())
+      if (ethTransaction.getNonce.subtract(stateNonce).compareTo(MaxNonceGap) >= 0) {
+        log.trace(s"Transaction $ethTransaction nonce gap respect state nonce exceeds maximum allowed size: tx nonce ${ethTransaction.getNonce()}, " +
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:52]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala (hunk 1 above)
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:285]: Good catch! Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala (hunk 2 above)
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:382]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/exception/MempoolException.scala (hunk 3 above)
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/exception/MempoolException.scala:6]: I agree. Done

File: sdk/src/main/scala/com/horizen/certificatesubmitter/AbstractCertificateSubmitter.scala (hunk 4 above)
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/certificatesubmitter/AbstractCertificateSubmitter.scala:85]: Done

File: sdk/src/test/scala/com/horizen/account/performance/AccountSidechainNodeViewHolderPerfTest.scala (hunk 5 above)
Comment (saratnt) [reply sdk/src/test/scala/com/horizen/account/performance/AccountSidechainNodeViewHolderPerfTest.scala:242]: Done

File: qa/resources/template.conf [hunk 8]
```diff
@@ -90,4 +90,9 @@ sparkz {
     enabled = %(REMOTE_KEY_MANAGER_ENABLED)s
     address = "http://127.0.0.1:5000"
   }
+
+ accountMempool {
```
[the remark this answers is not in the document]
Comment (saratnt) [reply qa/resources/template.conf:94]: I agree

**Review by alsala — CHANGES_REQUESTED** [review]

File: qa/SidechainTestFramework/account/ac_chain_setup.py [hunk 9]
```diff
@@ -19,7 +19,8 @@ class AccountChainSetup(SidechainTestFramework):
     def __init__(self, API_KEY='Horizen', number_of_mc_nodes=1, number_of_sidechain_nodes=1,
                  withdrawalEpochLength=LARGE_WITHDRAWAL_EPOCH_LENGTH, forward_amount=100,
                  block_timestamp_rewind=DEFAULT_EVM_APP_GENESIS_TIMESTAMP_REWIND, forger_options=None,
-                 initial_private_keys=None, circuittype_override=None, remote_keys_manager_enabled=False):
+                 initial_private_keys=None, circuittype_override=None, remote_keys_manager_enabled=False,
+                 max_nonce_gap=16):
```
[the remark this answers is not in the document]
Comment (saratnt) [reply qa/SidechainTestFramework/account/ac_chain_setup.py:23]: This comment is applicable to other constants as wel, so I have done it directly in PR:https://github.com/HorizenOfficial/Sidechains-SDK/pull/744

File: qa/SidechainTestFramework/sc_boostrap_info.py [hunk 10]
```diff
@@ -127,13 +127,14 @@ def __init__(self,
                  forger_options=SCForgerConfiguration(),
                  mempool_max_size=300,
                  mempool_min_fee_rate=0,
+                 max_nonce_gap=16,
```
[the remark this answers is not in the document]
Comment (saratnt) [reply qa/SidechainTestFramework/sc_boostrap_info.py:130]: Done in PR:https://github.com/HorizenOfficial/Sidechains-SDK/pull/744 (see previous comment)

File: qa/sc_evm_mempool_invalid_txs.py [hunk 11]
```diff
@@ -211,6 +210,23 @@ def run_test(self):
         assert_equal(0, len(response["transactionIds"]),
                      "Transaction that creates a smart contract with empty data added to node 1 mempool")
 
+        # Test that a transaction with nonce gap too big is rejected by the mem pool
+        exception_occurs = False
+        try:
+            createEIP1559Transaction(sc_node_1, fromAddress=evm_address_sc1, toAddress=None,
+                                          nonce=100, gasLimit=10, maxPriorityFeePerGas=900000000,
```
[the remark this answers is not in the document]
Comment (saratnt) [reply qa/sc_evm_mempool_invalid_txs.py:217]: Good point. Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 12]
```diff
@@ -16,10 +17,13 @@ import scala.util.Try
 
 class MempoolMap(
                   accountStateReaderProvider: AccountStateReaderProvider,
-                  baseStateReaderProvider: BaseStateReaderProvider) extends SparkzLogging {
+                  baseStateReaderProvider: BaseStateReaderProvider,
+                  mempoolSettings: AccountMempoolSettings) extends SparkzLogging {
   type TxIdByNonceMap = mutable.SortedMap[BigInteger, ModifierId]
   type TxByNonceMap = mutable.SortedMap[BigInteger, SidechainTypes#SCAT]
 
+  val maxNonceGap: BigInteger = BigInteger.valueOf(mempoolSettings.maxNonceGap)
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:25]: Done

File: sdk/src/test/scala/com/horizen/account/mempool/AccountMemoryPoolTest.scala [hunk 13]
```diff
@@ -56,7 +59,7 @@ class AccountMemoryPoolTest
     assertEquals("Wrong tx list size ", 1, listOfExecTxs.size)
     assertEquals("Wrong tx ", account1ExecTransaction0.id(), listOfExecTxs.head.id)
 
-    val account1NonExecTransaction0 = createEIP1559Transaction(value, BigInteger.valueOf(1000), Option(account1Key))
+    val account1NonExecTransaction0 = createEIP1559Transaction(value, BigInteger.valueOf(15), Option(account1Key))
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/test/scala/com/horizen/account/mempool/AccountMemoryPoolTest.scala:62]: Done

**Review by i-Alex — APPROVED** [review]

**Review by alsala — APPROVED** [review]

**Pull request merged**